File: blackjack/matlab/partitions.m
function m = partitions(cards, subtotal)
  m=0;
  % Hit
  for i=1:1:10
    if (cards(i)>0)
      total=subtotal+i;
      if (total < 21)
        % Stand
        m=m+1;
        % Hit again
        cards(i)=cards(i)-1;
        m=m+partitions(cards, total);
        cards(i)=cards(i)+1;
      elseif (total==21)
        % Stand; hit again is an automatic bust
        m=m+1;
      end
    end
  end
end

File: blackjack/python-swift/partitions.swift
public func partitions(subtotal:Int,cards:UnsafeMutableBufferPointer<Int>) -> Int {
    func p(cards: inout [Int], subtotal: Int) -> Int {
        var m = 0
        // Hit
        for i in 0...9 {
            if (cards[i]>0) {
                let total = subtotal+i+1
                if (total < 21) {
                    // Stand
                    m += 1
                    // Hit again
                    cards[i] += -1
                    m += p(cards:&cards,subtotal:total)
                    cards[i] += 1
                } else if (total==21) {
                    // Stand; hit again is an automatic bust
                    m += 1
                }
            }
        }        
        return m
    }

    var a: Array<Int> = []
    for i in 0...9 {
        a.append(cards[i])
    }

    return p(cards:&a, subtotal:subtotal)
}
